src/parser.ts

```diff
@@ -0,0 +1,261 @@
+// [VexFlow](http://vexflow.com) - Copyright (c) Mohit Muthanna 2010.
+//
+// A generic text parsing class for VexFlow.
+
+import { Vex } from './vex';
+import { Grammar } from './easyscore';
+
+//////////////////////////////////////////////////////////////////////////////////////////////////
+// TODO: Move to easyscore.ts
+type RuleFunction = () => Rule;
+type TriggerFunction = (state?: Result) => void;
+interface Rule {
+  // Lexer Rules
+  token?: string; // The token property is a string that is compiled into a RegExp.
+  noSpace?: boolean; // TODO: None of the EasyScore rules specify noSpace, so it is not used anywhere.
+
+  // Parser Rules
+  expect?: RuleFunction[];
+  zeroOrMore?: boolean;
+  oneOrMore?: boolean;
+  maybe?: boolean;
+  or?: boolean;
+  run?: TriggerFunction;
+}
+// END: Move to easyscore.ts
+//////////////////////////////////////////////////////////////////////////////////////////////////
+
+// To enable logging for this class. Set `Vex.Flow.Parser.DEBUG` to `true`.
+function L(...args: any[]) {
+  if (Parser.DEBUG) Vex.L('Vex.Flow.Parser', args);
+}
+
+export const X = Vex.MakeException('ParserError');
+
+const NO_ERROR_POS = -1;
+
+interface Result {
+  success: boolean;
+
+  // Lexer Results
+  pos?: number;
+  incrementPos?: number;
+  matchedString?: string;
+
+  // Parser Results
+  matches?: Match[];
+  numMatches?: number;
+  results?: Result[];
+  errorPos?: number; // Set to NO_ERROR if successful. N if there is an error in the string.
+}
+
+// Converts parser results into an easy to reference list that can be used in triggers.
+// eventually returns: string (including empty strings) | nested array | null
+type Match = null | string | Array<Match>;
```

Comment:
We are moving from `null` to `undefined`. Change to:
```
type Match = string | Array<Match>
```
Would it be possible simply:
```
type Match = string | string[]
```
?

Reply:
Hmm... we should consult with @0xfe on this. I like the idea of being consistent throughout the code base, and if we are using `undefined` instead of `null` everywhere else, that's fine by me.

However, for this TS migration I tried to stick as much as possible to the original code's behavior. My only goal was to apply typings and enforce type checks to capture bugs at compile time. I didn't want to change the actual output of the method calls.

In the original parser, the `flattenMatches(...)` function returns a nested array in which the leaf elements are either string or null. This array is actually returned to an external class (e.g., Grammar in easyscore.js). The `run: (state) => ...` triggers take this input and calls code to build up the VexFlow objects. It is possible that someone 

I haven't looked deep enough to know if changing null to undefined will break anything in this case. It is possible that those builder functions check for null instead of undefined.

In any case, I'll wait for Mohit to to chime in before I start changing how the code behaves. Maybe this should be a separate PR that happens after the migration for parser.ts. That PR can tackle the conversion from `null` to `undefined` and make sure it doesn't break any external code.

Reply:
You are right:
```
type Match = string | string[]
```
makes no sense. The `undefined` was suggested to go for ? in parameters and attributes when they ar not mandatory.